### PR TITLE
Controller bootstrap

### PR DESCRIPTION
**Pull request by makhov**
CAPI controller for k0s control plane bootstrap. 
Doesn't support scaling controller nodes down atm.

### CODE REVIEW

File: internal/controller/controlplane/k0s_controlplane_controller.go [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+	}
+
+	res, err = c.reconcile(ctx, cluster, kcp)
+	if err != nil {
+		return res, err
+	}
+
+	// TODO: We need to have bit more detailed status and conditions handling
+	kcp.Status.Ready = true
+	kcp.Status.ExternalManagedControlPlane = true
```
Comment (jnummelin) [inline internal/controller/controlplane/k0s_controlplane_controller.go:108]: AFAIK this should be false when we manage CP in `Machine`s

File: internal/controller/bootstrap/controlplane_bootstrap_controller.go [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("error setting control plane endpoint: %v", err)
+		}
+
+		k0sConfigBytes, err := config.Spec.K0s.MarshalJSON()
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("error marshalling k0s config: %v", err)
+		}
+		files = append(files, cloudinit.File{
+			Path:        "/tmp/k0s.yaml",
```
Comment (jnummelin) [inline internal/controller/bootstrap/controlplane_bootstrap_controller.go:154]: Why on `/tmp`? These are `Machine`s and thus reboot would probably wipe out `/tmp` and render the controller bust.

Comment (makhov) [reply internal/controller/bootstrap/controlplane_bootstrap_controller.go]: Moved it to `/etc/`

File: internal/controller/controlplane/k0s_controlplane_controller.go (hunk 1 above)
Comment (makhov) [reply internal/controller/controlplane/k0s_controlplane_controller.go]: Done

File: internal/controller/bootstrap/controlplane_bootstrap_controller.go [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+			Namespace: "kube-system",
+		},
+		Type: corev1.SecretTypeBootstrapToken,
+		StringData: map[string]string{
+			"token-id":     tokenID,
+			"token-secret": tokenSecret,
+			// TODO We need bit shorter time for the token
+			"expiration":                       time.Now().Add(24 * time.Hour).Format(time.RFC3339),
+			"usage-bootstrap-api-auth":         "true",
+			"description":                      "Worker bootstrap token generated by k0smotron",
```
Comment (jnummelin) [inline internal/controller/bootstrap/controlplane_bootstrap_controller.go:349]: nit: Isn't this controller bootstrap token :)

File: internal/controller/bootstrap/controlplane_bootstrap_controller.go [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+
+func (c *ControlPlaneController) createKubeconfig(ctx context.Context, scope *Scope) error {
+	if scope.Cluster.Spec.ControlPlaneEndpoint.IsZero() {
+		return errors.New("control plane endpoint is not set")
+	}
+
+	return kubeconfig.CreateSecret(ctx, c.Client, scope.Cluster)
+}
+
+func (c *ControlPlaneController) getChildClusterClientSet(ctx context.Context, scope *Scope) (*kubernetes.Clientset, error) {
```
Comment (jnummelin) [inline internal/controller/bootstrap/controlplane_bootstrap_controller.go:364]: https://pkg.go.dev/sigs.k8s.io/cluster-api@v1.5.0/controllers/remote#NewClusterClient

There's a generic helper for this

**Review by jnummelin — COMMENTED** [review]
Couple minor nits only. good stuff

File: internal/controller/bootstrap/controlplane_bootstrap_controller.go (hunk 4 above)
Comment (makhov) [reply internal/controller/bootstrap/controlplane_bootstrap_controller.go]: Re-wrote it using the helper, thanks!

File: internal/controller/bootstrap/controlplane_bootstrap_controller.go (hunk 3 above)
Comment (makhov) [reply internal/controller/bootstrap/controlplane_bootstrap_controller.go]: Fixed!

**Review by jnummelin — APPROVED** [review]

**Pull request merged**